ludwig/utils/triton_utils.py

```diff
@@ -49,142 +89,488 @@ def forward(self, {input_signature}):
+        os.makedirs(self.base_path, exist_ok=True)
+
+        self.output_data_example: Dict[str, Union[TorchscriptPreprocessingInput, torch.Tensor]] = self.module(
+            self.input_data_example
+        )
+
+        # generate input and output features.
+        self.input_features: List[TritonConfigFeature] = []
+        for i, (feature_name, content) in enumerate(self.input_data_example.items()):
+            ludwig_type = "tensor"
+            # "block" is temporary
+            if self.inference_stage == PREPROCESSOR or self.inference_stage == "block":
+                ludwig_type = self.ludwig_config[INPUT_FEATURES][i][TYPE]
+            self.input_features.append(
+                TritonConfigFeature(feature_name, ludwig_type, content, self.inference_stage, INPUT, i)
+            )
+
+        self.output_features: List[TritonConfigFeature] = []
+        for i, (feature_name, content) in enumerate(self.output_data_example.items()):
+            ludwig_type = "tensor"
+            self.output_features.append(
+                TritonConfigFeature(feature_name, ludwig_type, content, self.inference_stage, OUTPUT, i)
             )
+
+    def save_model(self) -> str:
+        """Scripts the model and saves it."""
+        if not isinstance(self.model_version, int) or self.model_version < 1:
+            raise ValueError("Model version has to be a non-zero positive integer")
+        pass
+
+        os.makedirs(os.path.join(self.base_path, str(self.model_version)), exist_ok=True)
+        model_path = os.path.join(self.base_path, str(self.model_version), "model.pt")
+        self.model_ts = TritonModel(
+            self.module, self.input_features, self.output_features, self.inference_stage
+        ).generate_scripted_module()
+        self.model_ts.save(model_path)
+
+        # this is to debug
+        wrapper_definition = TritonModel(
+            self.module, self.input_features, self.output_features, self.inference_stage
+        ).generate_inference_module_wrapper()
+        with open(os.path.join(self.base_path, "wrapper.py"), "w") as f:
+            f.write(wrapper_definition)
+
+        return model_path
+
+    def save_config(self) -> str:
+        """Save the Triton config."""
+        self.config = TritonConfig(self.full_model_name, self.input_features, self.output_features)
+        config_path = os.path.join(self.base_path, "config.pbtxt")
+        with open(config_path, "w") as f:
+            f.write(self.config.get_model_config())
+        return config_path
+
+
+@dataclass
+class TritonEnsembleConfig:
+    """Dataclass for creating and saving the Triton ensemble config."""
+
+    triton_master_preprocessor: TritonMaster
+    triton_master_predictor: TritonMaster
+    triton_master_postprocessor: TritonMaster
+    model_name: str
+    output_path: str
+    model_version: int
+
+    def __post_init__(self):
+        self.ensemble_model_name = self.model_name + "_" + ENSEMBLE
```

Comment:
Given this is the `entrypoint` for the model, and we will need a way to reference this upstream, I suggest not having a suffix for the ensemble file such that if the `model_name` was `titanic_ensemble` then the directory structure would be the following.  

```
model_repository
├── titanic_ensemble
│   ├── 1
│   │   └── model.txt
│   ├── 2
│   │   └── model.txt
│   └── config.pbtxt
├── titanic_ensemble_postprocess
│   ├── 1
│   │   └── model.pt
│   ├── 2
│   │   └── model.pt
│   └── config.pbtxt
├── titanic_ensemble_predict
│   ├── 1
│   │   └── model.pt
│   ├── 2
│   │   └── model.pt
│   └── config.pbtxt
└── titanic_ensemble_preprocess
    ├── 1
    │   └── model.pt
    ├── 2
    │   └── model.pt
    └── config.pbtxt
```

Also, for your constants I notice these strings are `UPPERCASE` perhaps these should be lower?